packages/apollo-engine-reporting/src/extension.ts

```diff
@@ -455,3 +428,153 @@ function defaultGenerateClientInfo({ request }: GraphQLRequestContext) {
+      // Custom function to allow user to specify what variablesJson will look like
+      const originalKeys = Object.keys(variables);
+      const modifiedVariables = sendVariableValues.valueModifier({
+        variables: variables,
+        operationString: operationString,
+      });
+      return cleanModifiedVariables(originalKeys, modifiedVariables);
+    } else {
+      return variables;
+    }
+  })();
+
+  // Note: we explicitly do *not* include the details.rawQuery field. The
+  // Engine web app currently does nothing with this other than store it in
+  // the database and offer it up via its GraphQL API, and sending it means
+  // that using calculateSignature to hide sensitive data in the query
+  // string is ineffective.
+  Object.keys(variablesToRecord).forEach(name => {
+    if (
+      sendVariableValues == null ||
+      ('safelistAll' in sendVariableValues &&
+        !sendVariableValues.safelistAll) ||
+      ('exceptVariableNames' in sendVariableValues &&
+        // We assume that most users will have only a few private variables,
+        // or will just set privateVariables to true; we can change this
+        // linear-time operation if it causes real performance issues.
+        sendVariableValues.exceptVariableNames.includes(name))
+    ) {
+      // Special case for private variables. Note that this is a different
+      // representation from a variable containing the empty string, as that
+      // will be sent as '""'.
+      details.variablesJson![name] = '';
+    } else {
+      try {
+        details.variablesJson![name] =
+          variablesToRecord[name] == null
+            ? ''
+            : JSON.stringify(variablesToRecord[name]);
+      } catch (e) {
+        // This probably means that the value contains a circular reference,
+        // causing `JSON.stringify()` to throw a TypeError:
+        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#Issue_with_JSON.stringify()_when_serializing_circular_references
+        details.variablesJson![name] = JSON.stringify(
+          '[Unable to convert value to JSON]',
+        );
+      }
+    }
+  });
+
+  return details;
+}
+
+// Creates trace details when privateVariables [DEPRECATED] was set.
+// Wraps privateVariable into a format that can be passed into makeTraceDetails(),
+// which handles the new option sendVariableValues.
+export function makeTraceDetailsLegacy(
+  variables: Record<string, any>,
+  privateVariables: Array<String> | boolean,
+): Trace.Details {
+  const newArguments = Array.isArray(privateVariables)
+    ? {
+        exceptVariableNames: privateVariables,
+      }
+    : {
+        safelistAll: !privateVariables,
+      };
+  return makeTraceDetails(variables, newArguments);
+}
+
+// Helper for makeTraceDetails() to enforce that the keys of a modified 'variables'
+// matches that of the original 'variables'
+function cleanModifiedVariables(
```

Comment:
I dunno that this is particularly necessary (if people want to send us weird data, they can send us weird data) but I'm not opposed to it.

Reply:
@glasser @zionts It would actually be really nice to be able to add additional data to the variables. In my case I'm looking to add a userId so that I can identify particular users inside of apollo engine's reporting interface.